### ISSUE TITLE
Compiling problems Debian Sarge (32bit)

### ISSUE THREAD

**Issue by robbo007** [kej715/DtCyber#4]
Hi, I'm trying to compile under Debian Sarge and I get the following error:

CYBER:/cyber/DtCyber-main# make -f Makefile.linux32 dtcyber
cc -02 -I. -I/usr/X11R6/include -std=gnu99 -c operator.c
operator.c: In function 'opWait':
operator.c:1886: error: useconds t'undeclared (first use in this function)
operator.c:1086: error: (Each undeclared identifier is reported only once
operator.c:1086: error: for each function it appears in.)
make: *** [operator.o] Error 1
CYBER:/cuber/DtCyber-main#

Any ideas?

**Comment by kej715** [comment]
Hi,

"useconds_t" is defined in unistd.h, and is defined by the usleep() system
service. Debian Sarge is quite ancient (17+ years old), and it has been
archived. More contemporary versions of Debian will have usleep() and
useconds_t. Is it possible for you to upgrade?

cheers,
Kevin

**Comment by robbo007** [comment]
Hi Kevin,
Unfortunately I do as its the only version that seems to work with my hardware. Is there a work around for this?

**Comment by kej715** [comment]
Hi,

The operator.c module needs to be able to sleep for a number of
milliseconds. usleep() accomplishes this on Linux (except, apparently, old
versions of Debian), and Sleep() accomplishes it on Windows. See this Stack
Overflow article about sleeping for milliseconds in Linux:


https://stackoverflow.com/questions/1157209/is-there-an-alternative-sleep-function-in-c-to-milliseconds

The article provides a workaround for systems that do not have usleep() or
nanosleep() -- it suggests using select() with all bit vectors NULL. The
article also mentions that usleep() has been deprecated, and nanosleep()
should be used instead, so DtCyber should be updated to use nanosleep
instead of usleep. Unfortunately, that probably won't work on Debian Sarge
either, so using select() is probably the best (and only) workaround.

cheers,
Kevin

**Comment by wschaub** [comment]
I pulled out a copy of sarge via the magic of docker and played around, it looks like what we need is hidden behind feature macro tests. 

Try replacing the CFLAGS line in your Makefile with this and tell me how it goes, I got it to compile with the latest github checkout with this in the sarge docker container but I haven't tested the resulting binary.

CFLAGS  = -O2 -D_GNU_SOURCE  -I. $(INCL) -std=gnu99

**Comment by robbo007** [comment]
@wschaub Amazing thanks :) That seemed to compile ok. I forgot to install the libx11-dev package as well so I added them. I now need to configure and see if I can get it loading. I'll keep you posted. Regards.

**Comment by wschaub** [comment]
I just did another test on the sarge container, this time to test if the ancient glibc supports nanosleep. looks like it does, and without any need to set a define, pulling in <time.h> will get you the definition of timespec and nanosleep.

**Comment by kej715** [comment]
Thanks William!  If nanosleep is supported in the ancient version of
Debian, then that makes replacing usleep by nanosleep even more compelling.
I'll set to work on that right away.

cheers,
Kevin

**Comment by wschaub** [comment]
I went through and tried to get everything working possible on sarge, I got enough working to run the automation scripts to build NOS1.3 and NOS2.8.7 but I wasn't able to get git-lfs working because Go does not support 2.6 and below kernels anymore. 

I uploaded the updated curl, openssl, git, tcl and expect  here https://steubentech.com/sarge-tools/

**Comment by robbo007** [comment]
Hi @wschaub, 
Does this mean NOS won't run then. Not having support for git-lfs-3.2.0 and go1.18.2? Should I concentrate getting my hardware to work with maybe Debian Lenny?

**Comment by robbo007** [comment]
@wschaub ok thanks for clarifying that. I've manged to get it compiled. I have all the files that Tom Hunter gave me back in the day. 

dtcyber loads but then dumps back to Debian with a "could not open display" error.  Could this be because Xwindows is holding the resources? 

```
Desktop CYBER 5.6.1 - Copyright (C) Tom Hunter
Licensed under the terms of the GNU General Public License version 3

Starting initialisation
Unexpected length of CM backing file, clearing CM
CPU model CYBER865 initialised (CM: 4000000, ECS: 0)
Unexpected length of PPM backing file, clearing PPM
PPs initialised (number of PPUs 24)
Channels initialised (number of channels 40)
Using gettimeofday() clock at 1.000000 MHz
Status/Control Register initialised on channel 16
Status/Control Register initialised on channel 36
Disk with 843 cylinders initialised on channel 1 unit 0
Disk with 843 cylinders initialised on channel 2 unit 0
Disk with 843 cylinders initialised on channel 3 unit 0
Disk with 843 cylinders initialised on channel 4 unit 0
Disk with 843 cylinders initialised on channel 1 unit 1
Disk with 843 cylinders initialised on channel 2 unit 1
Disk with 843 cylinders initialised on channel 3 unit 1
Disk with 843 cylinders initialised on channel 4 unit 1
Disk with 843 cylinders initialised on channel 1 unit 2
Disk with 843 cylinders initialised on channel 2 unit 2
Disk with 843 cylinders initialised on channel 3 unit 2
Disk with 843 cylinders initialised on channel 4 unit 2
Console initialised on channel 10
Equipment 07, Unit 00 attached to DCC6681 on channel 7
LP3555/512 initialised on channel 7 equipment 7
Equipment 07, Unit 00 attached to DCC6681 on channel 11
CR3447 initialised on channel 11 equipment 7
Equipment 06, Unit 00 attached to DCC6681 on channel 11
CP3446 initialised on channel 11 equipment 6
MT679 initialised on channel 13 equipment 0 unit 0
MT679 initialised on channel 13 equipment 0 unit 1
MT679 initialised on channel 13 equipment 0 unit 2
MT679 initialised on channel 13 equipment 0 unit 3
Two port MUX initialised on channel 15
NPU initialised on channel 5 equipment 7
       Host ID: CYBER
  Coupler node: 1
      NPU node: 2

Desktop CYBER 5.6.1 - Copyright (C) Tom Hunter.
Licensed under the terms of the GNU General Public License version 3.
For details see included text file 'license-gpl-3.0.txt' or visit
'http://www.gnu.org/licenses'.

Operator interface
Please enter 'help' to get a list of commands

Operator>/cyber/CybisRelease1#
```

**Comment by robbo007** [comment]
I'm running everything on the same machine and running via the console. I've done export DISPLAY=0.0 
Does this require a certain resolution? I've tested my Xwindows and it loads fine using a maximum of 800x600 res. 
Regards
